### ISSUE TITLE
Killing the debugger leaves terminal in bad state

### ISSUE THREAD

**Issue by antocuni** [pdbpp/pdbpp#28]
Originally reported by: **dwt (Bitbucket: [dwt](https://bitbucket.org/dwt), GitHub: [dwt](https://github.com/dwt))**

----------------------------------------

I'm using pdbpp in web apps that auto restart when I change one of their files. This regularly leaves my terminal in a unusable state if the restart happens while the debugger is stopped somewhere.

This regularly happens to me because I stop the debugger, gain some information and then change the code to reflect that knowledge, just to find that I just garbled my terminal because I forgot to continue the debugger before saving my changes - which then kills the debugger as the web app restarts. :/

Please fix this: pdbpp should always reset terminal state if it is killed.

----------------------------------------
- Bitbucket: https://bitbucket.org/antocuni/pdb/issue/28

**Comment by antocuni** [comment]
*Original comment by* **Martin Matusiak (Bitbucket: [numerodix](https://bitbucket.org/numerodix), GitHub: [numerodix](https://github.com/numerodix))**:

----------------------------------------

This is actually the case in classic pdb too (I have the same use case). What I do a lot is develop a web app in say django, which includes a reloader that detects changes to any modules and restarts the process when that occurs. This will terminate pdb and leave the terminal in a state where input from the keyboard is not being echoed.

This should perhaps be fixed in pdb itself by catching the signal and restoring the terminal before exiting.

**Comment by antocuni** [comment]
*Original comment by* **dwt (Bitbucket: [dwt](https://bitbucket.org/dwt), GitHub: [dwt](https://github.com/dwt))**:

----------------------------------------

I'm not familiar with the internals of pdbpp - would fixing this in pdb fix it for pdbpp as well? 

It is pretty annoying, especially because it sometimes confuses my terminal so hard, that I can't provide input to it anymore until I SIGSTOP the current process from a different terminal and then  run a 'reset'. :/

**Comment by antocuni** [comment]
*Original comment by* **Antonio Cuni (Bitbucket: [antocuni](https://bitbucket.org/antocuni), GitHub: [antocuni](https://github.com/antocuni))**:

----------------------------------------

I cannot reproduce the behavior:
```
homer ~ $ python
Python 2.7.4 (default, Sep 26 2013, 03:20:26) 
[GCC 4.7.3] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> import pdb
>>> pdb.set_trace()
--Return--
[7] > <console>(1)<module>()->None
(Pdb++) import os
(Pdb++) os.kill(os.getpid(), 9)
Killed
homer ~ $ 
```

after the kill, my terminal looks perfectly fine. However, note that most probably what you describe is due to pyrepl, not pdbpp and thus should be reported there.

**Comment by antocuni** [comment]
*Original comment by* **Martin Matusiak (Bitbucket: [numerodix](https://bitbucket.org/numerodix), GitHub: [numerodix](https://github.com/numerodix))**:

----------------------------------------

I tried this with django just now and I couldn't reproduce it. I have this problem in flask, and I suspect it has more to do with how flask works than pdb.

Filed a bug: https://github.com/mitsuhiko/flask/issues/1076

Isolated test case: https://gist.github.com/numerodix/c13ce7c1c0917b3b7dd9

**Comment by AlexTelon** [comment]
I am getting this error during development. I tried to do a minimal example that triggers the bug.

```
import _thread
import time

def print_time( threadName, delay):
   while True:
       print ("hi")
       # a set_trace in code must be done.
       import pdb; pdb.set_trace()
       time.sleep(delay)

if __name__ == '__main__':
    # only one thread is needed
    _thread.start_new_thread( print_time, ("Thread-1", 1))

    # keep alive
    while True:
        pass
```

If I run this code:

`python test.py`

wait until I get a printout of the lines around the breakpoint (I get one after a few sec as I think I have sticky by default). Once you see a `sticky` prompt or maybe an `ll` do `Ctrl-c`

This causes the invisible text bug.

Doing `Ctrl-c` quickly before the `ll`-type prompt shows up does **not** cause the bug.
I also only get this issue if I have threads.

Also it also seems I must use the `pdb.set_trace() ` within the threads loop. Placing the `pdb.set_trace()` outside the loop in  `print_time` function and simply setting a breakpoint at line 6 does not reproduce this issue.

In other words the code below does not have an issue. If you run this and manually set a breakpoint at line 6 then `Ctlr-c` does not exit the program but `q` does. So maybe this case does not "fix" the issue but rather it forces the user to quit with `q` instead. Using `q` never gives me this issue. But sometimes I am not able to quit, which happens when I have a `set_trace()` within threads that do something in a loop. In those cases I am forced to do `Ctrl-c` and that causes the bug.
```
def print_time( threadName, delay):
    import pdb; pdb.set_trace()
    while True:
        print ("hi")
        time.sleep(delay)
```

Also:

The same issue happens when using:

`import pdb; pdb.pdb.set_trace()`

So the earlier comments seem to be right regarding this being a bug from pdb

**Comment by blueyed** [comment]
@AlexTelon 
Thanks for the report, I'll try to investigate/reproduce it later.
Does it also happen without pdbpp for you?
(A quick thought was that it might be connected to stdin being connected properly, but that should not be an issue with threads.)

**Comment by blueyed** [comment]
From https://github.com/antocuni/pdb/issues/28#issuecomment-307630744 it sounds like it might be an issue with pdb itself.
I could imagine pdbpp working around it, but it should really get fixed in cpython/pdb then.

**Comment by blueyed** [comment]
Here is a test case to reproduce this (also with pdb, Python 3.7.1).

It starts a Python process in the background, where pdb is triggered, kills it,
and then starts it again:

```
#!/usr/bin/env bash
set -x
python -c '__import__("pdb").set_trace()' </dev/tty &
pid=$!

sleep 1
kill "$pid"

python -c '__import__("pdb").set_trace()'
```

A fix / workaround is to use `reset -I` before the last command.
